mergify_engine/eventlogs.py

```diff
@@ -0,0 +1,170 @@
+# -*- encoding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import datetime
+import typing
+
+import daiquiri
+import msgpack
+
+from mergify_engine import date
+from mergify_engine import github_types
+from mergify_engine.dashboard import subscription
+
+
+if typing.TYPE_CHECKING:
+    from mergify_engine import context
+
+LOG = daiquiri.getLogger(__name__)
+
+DEFAULT_VERSION = 1
+EVENTLOGS_EXPIRATION = datetime.timedelta(days=90)
+EVENTLOGS_MAX_SIZE = 100000
+
+
+class EventBase(typing.TypedDict):
+    timestamp: datetime.datetime
+
+
+class EventMetadata(typing.TypedDict):
+    pass
+
+
+class EventLabelMetadata(EventMetadata):
+    added: typing.List[str]
+    removed: typing.List[str]
+
+
+class EventLabel(EventBase):
+    event_type: typing.Literal["label"]
+    metadata: EventLabelMetadata
+
+
+class EventRefreshMetadata(EventMetadata):
+    pass
+
+
+class EventRefresh(EventBase):
+    event_type: typing.Literal["refresh"]
+    metadata: EventRefreshMetadata
+
+
+def _get_pull_request_key(
+    owner_id: github_types.GitHubAccountIdType,
+    repo_id: github_types.GitHubRepositoryIdType,
+    pull_request: github_types.GitHubPullRequestNumber,
+) -> str:
+    return f"eventlogs/{owner_id}/{repo_id}/{pull_request}"
+
+
+EventName = typing.Literal["label", "refresh"]
+Event = typing.Union[EventLabel, EventRefresh]
+
+
+class GenericEvent(EventBase):
+    event_type: EventName
+    metadata: EventMetadata
+
+
+@typing.overload
+async def write(
+    repository: "context.Repository",
+    pull_request: github_types.GitHubPullRequestNumber,
+    event_type: typing.Literal["refresh"],
+    metadata: EventRefreshMetadata,
+) -> None:
+    ...
+
+
+@typing.overload
+async def write(
+    repository: "context.Repository",
+    pull_request: github_types.GitHubPullRequestNumber,
+    event_type: typing.Literal["label"],
+    metadata: EventLabelMetadata,
+) -> None:
+    ...
```

Comment:
```suggestion
@typing.overload
async def write(
    repository: "context.Repository",
    pull_request: github_types.GitHubPullRequestNumber,
    event_type: typing.Union[typing.Literal["refresh"],typing.Literal["label"]]
    metadata: EventRefreshMetadata,
) -> None:
    ...
```
is it better to merge the two @overload definitions or keep them separate ?